### ISSUE TITLE
Encryption at rest key rotation is not working

### ISSUE THREAD

**Issue by nopa12** [kubernetes/kubernetes#49565]
/kind bug

**What happened**:
While following [Encryption at Rest guide](https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/), the encryption key rotation isn't working. The secrets are still encrypted with the old key.

**What you expected to happen**:
The secrets will be encrypted with the new key.

**How to reproduce it (as minimally and precisely as possible)**:
Just follow the guide.

**Anything else we need to know?**:
The cluster was installed with kubeadm v1.7.1 .
The same issue happens when trying to remove the encryption and decrypt the secrets with the actions at the end of the guide. After disabling the encryption and putting 'identity: {}' as the first encryption provider in the EncryptionConfig file the secrets still encrypted inside etcd with the old encryption key.

**Environment**:
- Kubernetes version: client 1.7.1, server 1.7.2
- OS: ubuntu 16.04.02 desktop
- Kernel: Linux laptop 4.10.0-27-generic #30~16.04.2-Ubuntu SMP Thu Jun 29 16:07:46 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
- Install tools: kubeadm v1.7.1

**Comment by nopa12** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by php-coder** [comment]
CC @smarterclayton @liggitt

**Comment by liggitt** [comment]
```
kubectl get secrets -o json | kubectl replace -f -
```

that command would only modify secrets in a single namespace... did you check in that namespace, or did you mean to perform the decryption across all namespaces (adding `--all-namespaces`)?

**Comment by nopa12** [comment]
I didn't mean to perform encryption on all namespaces. Only on my keys.

I will provide more info. I have a secret, `personal-secret`, which is encrypted with `key1`. Now I changed my encryption configuration file, added `key2` as the first key:
```
kind: EncryptionConfig
apiVersion: v1
resources:`
  - resources:
    - secrets
    providers:
    - aescbc:
        keys:
        - name: key2
          secret: zwGFDIi3rznj50OLLOcb2+zwnPExZwC3k64Evzc/dnw=
        - name: key1
          secret: dFQ/o0NPFExWgOjt7KmgP0OZiFst9dPZnPKfz1gZxM8=
    - identity: {}
```

I restarted the apiserver and then executed:
```
$ kubectl get secrets -o json | kubectl replace -f -
secret "default-token-l69pr" replaced
secret "personal-secret" replaced
```
The expected result, according to the doc, is that the secret will be now encrypted with `key2`. But the secret is still encrypted with the old key:
```
$ docker exec -it 73d95cb5e73f /bin/sh -c 'ETCDCTL_API=3 etcdctl get /registry/secrets/default/personal-secret | hexdump -C'
00000000  2f 72 65 67 69 73 74 72  79 2f 73 65 63 72 65 74  |/registry/secret|
00000010  73 2f 64 65 66 61 75 6c  74 2f 70 65 72 73 6f 6e  |s/default/person|
00000020  61 6c 2d 73 65 63 72 65  74 0a 6b 38 73 3a 65 6e  |al-secret.k8s:en|
00000030  63 3a 61 65 73 63 62 63  3a 76 31 3a 6b 65 79 31  |c:aescbc:v1:key1|
00000040  3a 05 bb 94 f8 74 a0 81  ea 61 ad aa 56 03 e2 02  |:....t...a..V...|
00000050  e0 b8 8d 6d b3 fe 8c 33  19 4e f4 5e 2c c7 bb ba  |...m...3.N.^,...|
00000060  3b 11 30 2a 05 5f b1 04  f4 e2 3a a3 13 ca 9d 81  |;.0*._....:.....|
00000070  bd 38 e0 fc 36 42 46 b6  4a 0c 1c 9f 39 3e 32 a9  |.8..6BF.J...9>2.|
00000080  17 45 1d 33 9d 86 b2 86  3f ac b6 17 1c a8 45 8e  |.E.3....?.....E.|
00000090  84 10 6f 6d 60 1c 54 ae  c6 df 33 e3 66 de 43 7b  |..om`.T...3.f.C{|
000000a0  0f 3d 68 83 2f 86 4b 13  f7 2f e8 9a 0b 43 65 d0  |.=h./.K../...Ce.|
000000b0  4b 29 60 a2 5e 00 db 42  97 3a a3 72 e9 f2 bb 1c  |K)`.^..B.:.r....|
000000c0  9d ca ca 2b 4d 9e 51 b6  07 f0 db 7a 51 73 65 f3  |...+M.Q....zQse.|
000000d0  4e 2e 33 84 09 89 6b 59  20 68 87 6a d1 59 8a f1  |N.3...kY h.j.Y..|
000000e0  ac 4d d7 f2 fa 54 5f 06  ee 53 63 d7 7e eb 3b e4  |.M...T_..Sc.~.;.|
000000f0  43 c8 3d c6 9a a9 85 a1  b5 cc 57 91 20 85 64 c4  |C.=.......W. .d.|
00000100  ff 48 b7 cd 1b 30 ad a5  10 06 13 bf 35 7a 5f 38  |.H...0......5z_8|
00000110  01 be a2 a8 45 c2 18 99  36 51 e5 61 95 11 de 5c  |....E...6Q.a...\|
00000120  6b 64 1f dd b3 69 f3 73  0e 3b ac 09 c7 7d 8e 41  |kd...i.s.;...}.A|
00000130  63 41 db 07 5a 85 4a d9  c9 6f c3 3f 3d fb 36 68  |cA..Z.J..o.?=.6h|
00000140  a4 0a                                             |..|
00000142
```

**Comment by php-coder** [comment]
@nopa12 Is this issue still here?

**Comment by liggitt** [comment]
this was fixed in 1.7.8 and is working in 1.7.x

a separate bug was introduced and is currently affecting 1.8.4+ and 1.9+ in the same way

fix is in #58375
